store/tikv/coprocessor.go

```diff
@@ -1119,3 +1152,48 @@ func (it copErrorResponse) Next(ctx context.Context) (kv.ResultSubset, error) {
 func (it copErrorResponse) Close() error {
 	return nil
 }
+
+// EndCopWorkerAction implements memory.ActionOnExceed for copIteratorWorker. If
+// the memory quota of a query is exceeded, EndCopWorkAction.Action would end one copIteratorWorker.
+// If there is only one or zero worker is running, delegate to the fallback action.
+type EndCopWorkerAction struct {
+	once           sync.Once
+	fallbackAction memory.ActionOnExceed
+	mu             struct {
+		sync.Mutex
+		// exceeded indicates that datasource have exceeded memQuota.
+		exceeded uint32
+
+		// alive worker indicates how many copIteratorWorker are running
+		aliveWorker int
+	}
+}
+
+// Action sends a signal to trigger end one copIterator worker.
+func (e *EndCopWorkerAction) Action(t *memory.Tracker) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	// only one or zero worker is running, delegate to the fallback action
+	if e.mu.aliveWorker < 2 {
+		if e.fallbackAction != nil {
+			e.fallbackAction.Action(t)
+		}
+		return
+	}
+	// set exceeded as 1
+	e.once.Do(func() {
+		e.mu.exceeded = 1
+		logutil.BgLogger().Info("memory exceeds quota, end one copIterator worker.",
```

Comment:
not just shrink a thread, it also need to consider expand a new thread if the less workers cannot produce enough chunks after ending some.

Reply:
I think this is a good idea. Do we need realize this feature in this request or we create a new issue to track this feature?

Reply:
The drawback of just shrinking threads, it may decrease performance if the parent operators run faster. It is better to consider to consider expanding threads. To shrink threads, we may not end them, just suspend them, then resume them.

Reply:
Yes, we need to ensure that only one worker would be stopped or suspended after oom action.

Reply:
Whether suspend or stop a worker is ok to me as this request is going to realize the oom action for the cop worker, so I think it is unnecessary to talk about the how and when to reuse the cop worker in this request if we decided to suspend a worker instead of stopping it.